homeworks/week3/Minh-Le.java

```diff
@@ -0,0 +1,139 @@
+//Repl link: https://repl.it/@MinhD1/SinglyLinkedList#Main.java
+
+// Problem: Write two SinglyLinkedList operations: add and remove.
+public class Main
+{
+	public static void main(String[] args) {
+    SinglyLinkedList linkedList = new SinglyLinkedList();
+
+    // move from empty list
+    System.out.println(linkedList.remove(new Node(1))); // false
+
+    linkedList.print(); // empty
+    linkedList.add(0, new Node(0));
+    linkedList.print(); // 0->null
+    linkedList.add(1, new Node(1));
+    linkedList.print(); // 0->1->null
+    linkedList.add(2, new Node(2));
+    linkedList.print(); // 0->1->2->null
+
+    // add to the middle of the list
+    linkedList.add(1, new Node(3));
+    linkedList.print(); // 0->3->1->2->null
+
+    //index out of range, do nothing
+    linkedList.add(5, new Node(4));
+    linkedList.print(); // 0->3->1->2->null
+
+    // remove element that is not found
+    System.out.println(linkedList.remove(new Node(4))); // false
+    linkedList.print(); // 0->3->1->2->null
+
+    //remove last element
+    System.out.println(linkedList.remove(new Node(2))); // true
+    linkedList.print(); // 0->3->1->null
+
+    //remove element in the middle
+    System.out.println(linkedList.remove(new Node(3))); // true
+    linkedList.print(); // 0->1->null
+
+    //remove first element
+    System.out.println(linkedList.remove(new Node(0))); // true
+    linkedList.print(); // 1->null
+
+     //remove the only element
+    System.out.println(linkedList.remove(new Node(1))); // true
+    linkedList.print(); // empty
+	}
+    
+  static class SinglyLinkedList {
+
+    private Node head = null;
+      
+    // Add to a specific position; do nothing if the index is out of range or e is null
+    // Time complexity: worst case O(n), best case O(1), space complexity: O(1)
+    public void add(int index, Node e) {
+      if (e == null) return;
+
+      if (this.head == null && index == 0) {
+        this.head = e;
+        return;
+      }
+
+      int count = 0;
+      Node prev = null;
+      Node curr = this.head;
+      for (int i = 0; i < index && curr != null; i++) {
+        prev = curr;
+        curr = curr.next;
+        count++;
+      }
+
+      if (count < index) { // index out of range, do nothing
+        return; 
+      }
+
+      e.next = curr;
```

Comment:
Do you think you can make the code shorter and simpler if `curr` represents the node before `e`? 
You could get rid of the `prev` variable and the check in lines 58-61.

Reply:
Thank you for your valuable feedback. What you suggested indeed made the code shorter. I made some changes based on your suggestion.